src/gallia/services/uds/ecu.py

```diff
@@ -416,13 +395,18 @@ async def start_cyclic_tester_present(self, interval: float) -> None:
         await asyncio.sleep(0)
 
     async def stop_cyclic_tester_present(self) -> None:
+        logger.debug("Stopping tester present worker")
         if self.tester_present_task is None:
             logger.warning(
                 "BUG: stop_cyclic_tester_present() called but no task running"
             )
             return
 
-        self.tester_present_task.cancel()
+        running = True
```

Comment:
In my opinion, this commit fbc572333adb7ae131911f425ed755029925cd62 is not a proper fix for this not well understood problem of cancelling async tasks. It is still not clear why this problem actually occurs and if there are problems in the `gallia` codebase or elsewhere. So I would rather investigate the root cause of this. Maybe we should ask the Python matrix channel or on the relevant mailing lists before merging this.

Reply:
Feel free to investigate!

Nonetheless I suggest we merge this to avoid the issue. As a compromise, we could add a `#FIXME` or the like...

Reply:
You are free to investigate, too.

> I suggest we merge this to avoid the issue.

As far as I understand, this could create an infinite loop when the tasks decides to ignore cancelling, as it obviously does for some reason. Additionally, it could create an unexpected delay.

Perhaps we could avoid the cancelling problem at all by using something like this:

``` python
# Mutexes or similar avoided for readability in this example.

# In tester present task
    # …
    while self.running:
        do_stuff()

# In main task
def stop_tester_present(self):
    self.running = False
```

If synchronization is required, there is [asyncio.Event](https://docs.python.org/3/library/asyncio-sync.html#asyncio.Event) which could be used to wait until the worker has been terminated.

Reply:
I did investigate, and this is what I came up with. If you want to dig deeper you of course can...

If you prefer the solution with a class variable that could work I guess, but keep in mind that it might be an overly complex workaround for a bug in ansyncio that might be fixed at some point?

Regarding your two points:

1. In theory this could be a infinite loop, yes. In practice it never went higher than 3 iterations, so I would take the chance. Keep in mind, Keep in mind that the task not being cancelled is a very rare/racy bug.
2. What do you mean by unexpected delay? For whom exactly and where would that be a problem?

Reply:
Yeah, I am still not sure what is the best solution for this. I will have a look at this https://superfastpython.com/asyncio-task-cancellation-best-practices; maybe it helps.

Regarding points:

1. Yes, rare on one particular setup. Since nobody knows why and when this happens, it could behave differently on a different test setup.
2. When it takes more iterations to cancel the task, the scanner is blocked and it causes a delay that is not related to actual network traffic. This could lead to a lost UDS session, for instance. Not sure how realistic this scenario is.

I would like to test a few things on my machine first, in order to form a view on this problem.

Reply:
See b45ecd39c8c6dc2f34883edcbd2f237a28126ac6

Looks like a valid solution that avoids the `while True` approach.

I'm pretty sure the `asyncio.CancelledError` gets lost/consumed somewhere but no idea where and how.

Reply:
I understand you are OK with the changes in fbcea4310c1184139a0fcaffce2a25a0e2169aba ?

Care to resolve the conversation?